File: grpc/src/main/java/grpc/lib/compiler/GlobalDeclarations.java
package grpc.lib.compiler;

import grpc.lib.internal.GrpLexer;
import grpc.lib.internal.GrpParser;
import grpc.lib.symbol.*;
import org.antlr.v4.runtime.tree.TerminalNode;

import java.util.LinkedList;
import java.util.Optional;

public class GlobalDeclarations extends CompilerPhase {
    public Type getType(GrpParser.TypeContext ctx) {
        TerminalNode tn = (TerminalNode) ctx.getChild(0);
        int t = tn.getSymbol().getType();

        switch (t) {
            case GrpLexer.INT8:
                return Type.INT8;
            case GrpLexer.INT32:
                return Type.INT32;
            case GrpLexer.INT64:
                return Type.INT64;
            case GrpLexer.FLOAT:
                return Type.FLOAT;
            case GrpLexer.DOUBLE:
                return Type.DOUBLE;
            case GrpLexer.UINT8:
                return Type.UINT8;
            case GrpLexer.UINT32:
                return Type.UINT32;
            case GrpLexer.UINT64:
                return Type.UINT64;
            case GrpLexer.BOOL:
                return Type.BOOL;
            case GrpLexer.VOID:
                return Type.VOID;
            default:
                return Type.ERROR;
        }
    }

    public void redeclarationError(Location first, Location last, String name, SymType type) {
        String t = (type == SymType.FUNC)
                ? "function with same signature"
                : "global variable";

        String e = String.format("Redeclaration of %s `%s`. ", t, name);
        e += String.format("Previously declared at %s", last);

        addError(first, e);
    }

    @Override
    public void exitFdef(GrpParser.FdefContext ctx) {
        super.exitFdef(ctx);

        // Get the function info
        String name = ctx.Identifier().getText();
        Type rType = getType(ctx.type());
        Location location = new Location(ctx.Identifier());
        LinkedList<Variable> args = new LinkedList<>();

        // The argument list gets populated
        for (GrpParser.ArgContext arg : ctx.argList().arg()) {
            String argName = arg.Identifier().getText();
            Type argType = getType(arg.type());
            Location argLoc = new Location(arg.Identifier());
            Variable v = new Variable(argName, argType, name, argLoc);
            args.add(v);
        }

        Function function = new Function(name, rType, location, args);

        // Search for a function with the same name and signature
        Symbol[] qry = symbolTable.getSymbols(name, SymType.FUNC);
        Location l = null;

        if (qry.length > 0) {
            for (Symbol s : qry) {
                Function f = (Function) s;
                if (function.equals(f)) {
                    l = f.getLocation();
                    break;
                }
            }
        }

        if (l != null) {
            redeclarationError(location, l, name, SymType.FUNC);
        } else {
            // Both the functions and the args are added to the symbol table
            symbolTable.addSymbol(function);
            for (Variable v : args) {
                symbolTable.addSymbol(v);
            }
        }
    }

    @Override
    public void exitVdec(GrpParser.VdecContext ctx) {
        super.exitVdec(ctx);

        String name = ctx.Identifier().getText();
        Type type = getType(ctx.type());
        Location location = new Location(ctx.Identifier());

        Optional<Symbol> qry = symbolTable.getSymbol(name, "global", SymType.VAR);
        if (!qry.isPresent()) {
            Variable variable = new Variable(name, type, "global", location);
            symbolTable.addSymbol(variable);
        } else {
            Variable v = (Variable) qry.get();
            redeclarationError(location, v.getLocation(), name, SymType.VAR);
        }
    }
}

File: grpc/src/main/scala/grpc/lib/compiler/GlobalDeclarations.scala
package grpc.lib.compiler

import grpc.lib.internal.GrpParser.{FdefContext, SimpleStmtContext, VdecContext}
import grpc.lib.symbol.{Function, Location, SymType, Variable}

import scala.collection.mutable.ListBuffer

class GlobalDeclarations extends CompilerPhase {
  private var insideSimpleStmt = false

  /**
    * Reports a global function or variable redeclaration
    *
    * @param last  The location of the redeclaration
    * @param first The location of the previously declared symbol
    * @param name  The name of the symbol
    * @param typ  Whether it is a function or a variable
    */
  def redeclarationError(last: Location, first: Location, name: String,
                         typ: SymType.Value) = {
    val t = typ match {
      case SymType.FUNC => "function with same signature"
      case _ => "global variable"
    }

    val m = s"Redeclaration of $t `$name`. Previously declared at $first"
    addError(last, m.toString)
  }

  /**
    * Marks whenever the phase enters in a simple statement. Variables being
    * declared inside a simple statement are not global.
    *
    * @param ctx The context of the statement
    */
  override def enterSimpleStmt(ctx: SimpleStmtContext) = {
    super.enterSimpleStmt(ctx)
    insideSimpleStmt = true
  }

  /**
    * Marks whenever the phase leaves a simple statement. Variables being
    * declared inside a simple statement are not global
    *
    * @param ctx The context of the statement
    */
  override def exitSimpleStmt(ctx: SimpleStmtContext) = {
    super.exitSimpleStmt(ctx)
    insideSimpleStmt = false
  }

  /**
    * Inserts a function and its arguments into the symbol table
    *
    * @param ctx The context of the function definition
    */
  override def exitFdef(ctx: FdefContext) = {
    super.exitFdef(ctx)

    val name = ctx.Identifier().getText
    val returnType = getType(ctx.typ())
    val location = new Location(ctx.Identifier())
    val args = ListBuffer.empty[Variable]

    val ar = ctx.argList().arg()
    for (i <- 0 until ar.size()) {
      val argName = ar.get(i).Identifier().getText
      val argTyp = getType(ar.get(i).typ())
      val argLoc = new Location(ar.get(i).Identifier())
      val v = new Variable(argName, argTyp, name, argLoc)
      args += v
    }

    val function = new Function(name, returnType, location, args)

    val qry = symbolTable.getSymbol(name, SymType.FUNC)
    qry match {
      case Some(f) => redeclarationError(location, f.getLocation, name, SymType.FUNC)
      case _ => symbolTable.addSymbol(function)
    }
  }

  /**
    * Inserts a global variable into the symbol table
    *
    * @param ctx The context of the variable declaration
    */
  override def exitVdec(ctx: VdecContext) = {
    super.exitVdec(ctx)

    if (!insideSimpleStmt) {
      val name = ctx.Identifier().getText
      val typ = getType(ctx.typ())
      val location = new Location(ctx.Identifier())

      val variable = new Variable(name, typ, "global", location)

      val qry = symbolTable.getSymbol(name, "global", SymType.VAR)
      qry match {
        case Some(v) => redeclarationError(location, v.getLocation, name, SymType.VAR)
        case _ => symbolTable.addSymbol(variable)
      }
    }
  }
}
